### PR TITLE
Make CliPeon run shutdown hooks properly

### PR DESCRIPTION
**Pull request by drcrallen**

### CODE REVIEW

**Comment by fjy** [comment]
:+1:

File: services/src/main/java/io/druid/cli/CliPeon.java [hunk 1]
```diff
@@ -205,17 +205,35 @@ public void run()
...
         injector.getInstance(ExecutorLifecycle.class).join();
-        lifecycle.stop();
       }
       catch (Throwable t) {
         log.error(t, "Error when starting up.  Failing.");
         System.exit(1);
       }
+      log.info("Finished peon task");
+      try {
+        jettyServer.stop();
```
Comment (xvrl) [inline services/src/main/java/io/druid/cli/CliPeon.java]: is there a reason we need to shut down the jetty server separately? It should get stopped as part of the lifecycle.

**Comment by xvrl** [comment]
@drcrallen it should be enough to just make the Jetty query thread pool to use daemon threads, so we don't have to worry about it outside of the lifecycle.

**Comment by drcrallen** [comment]
Fixed jetty to have daemon threads and allow cleaner shutdown

File: services/src/main/java/io/druid/cli/CliPeon.java [hunk 2]
```diff
@@ -205,17 +205,29 @@ public void run()
   {
     try {
       Injector injector = makeInjector();
-
+      final Server jettyServer = injector.getInstance(Server.class);
```
Comment (xvrl) [inline services/src/main/java/io/druid/cli/CliPeon.java]: this is not needed anymore, is it?

File: services/src/main/java/io/druid/cli/CliPeon.java [hunk 3]
```diff
@@ -125,7 +124,8 @@ public void configure(Binder binder)
                                  .to(ServiceAnnouncingChatHandlerProvider.class).in(LazySingleton.class);
             handlerProviderBinder.addBinding("noop")
                                  .to(NoopChatHandlerProvider.class).in(LazySingleton.class);
-            binder.bind(ServiceAnnouncingChatHandlerProvider.class).in(LazySingleton.class);;
+            binder.bind(ServiceAnnouncingChatHandlerProvider.class).in(LazySingleton.class);
+            ;
```
Comment (gianm) [inline services/src/main/java/io/druid/cli/CliPeon.java]: extra semicolon

Comment (drcrallen) [inline services/src/main/java/io/druid/cli/CliPeon.java]: not mine! will fix anyways

Comment (xvrl) [inline services/src/main/java/io/druid/cli/CliPeon.java]: just for good measure :)

File: services/src/main/java/io/druid/cli/CliPeon.java [hunk 4]
```diff
@@ -205,17 +207,28 @@ public void run()
...
+                  public void run()
+                  {
+                    log.info("Running shutdown hook");
+                    lifecycle.stop();
+                  }
+                }
+            )
+        );
         injector.getInstance(ExecutorLifecycle.class).join();
-        lifecycle.stop();
```
Comment (gianm) [inline services/src/main/java/io/druid/cli/CliPeon.java]: I wonder if all of our regular task types are good about having every thread be non-daemon. If not, then I think they won't be able to shut down normally if you take out this lifecycle.stop().

Comment (gianm) [inline services/src/main/java/io/druid/cli/CliPeon.java]: As in, many of the tasks spawn a bunch of threads internally.

Comment (xvrl) [inline services/src/main/java/io/druid/cli/CliPeon.java]: @gianm are you saying we should have both the shutdown hook and the `lifecycle.stop()` after the lifecycle.join()?

Comment (gianm) [inline services/src/main/java/io/druid/cli/CliPeon.java]: I'm saying I'm not sure if it's needed or not, it depends on how good the existing tasks are about thread cleanliness on exit. If they're doing things right then the lifecycle.stop() is redundant to the shutdown hook.

@drcrallen suggested taking it out and letting the integration tests catch potential problems, which seems fair.

File: services/src/main/java/io/druid/cli/CliPeon.java (hunk 2 above)
Comment (drcrallen) [inline services/src/main/java/io/druid/cli/CliPeon.java]: nope fixed

File: services/src/main/java/io/druid/cli/CliPeon.java (hunk 1 above)
Comment (drcrallen) [inline services/src/main/java/io/druid/cli/CliPeon.java]: Was resolved in person. Related to daemon thread-ness of jetty thread pool.

**Comment by drcrallen** [comment]
Test failed on zk consistency issues. Restarting

**Comment by nishantmonu51** [comment]
:+1:

**Pull request merged**